feat: Allow Fileownership change through FSGroup and VOLUME_MOUNT_GROUP

mytreya-rh opened the pull request:
**What type of PR is this?**  
/kind feature

**What this PR does / why we need it**:  
(As of now pls consider it as a draft PR to discuss the solution further.)  
Allows the secrets to be mounted with FSGroup as specified in the POD spec.  
Thus, A pod with a non-root user should be able to read a secret, and that secret need not be world-readable.  

**Which issue(s) this PR fixes** :  
 Fixes #858 

**Is this a chart or deployment yaml update?**  
There is a yaml update for secrets-store.csi.x-k8s.io_secretproviderclasspodstatuses.yaml (generated through make manifests).  
It is added in the manifest_staging/deploy  
But if this PR merges after: https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/1622, the change in SecretProviderClassPodStatusStatus won't be required anymore and we can revert the changes related to reconciler.

**Special notes for your reviewer**:
### Problem:
  - As of now, all secrets mounted by ss-csi driver are with root:root ownership, and this is unwieldy because it would require much wider permissions on the secret, or elevated privileges on the workload
  - The standard way to set file ownership of volume mounts is through FSGroup.
  - However, that does not work with SS-CSI Driver even if FSGroupPolicy is set to "File" because:
      - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_mounter.go#L454
          - Here, ‘readonly’ volumes are skipped from being chown’d by csi_mounter even if the CSIDriver CRD is created with FSGroupPolicy of “File”
          - This part executes in the kubelet while it tries to provision the volume for the POD.
      - https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/pkg/secrets-store/nodeserver.go#L205
          - It is not possible to create a readwrite volume for ss-csi driver, and rightly so. Because, the secrets are only supposed to be read.
### Solution:
  - [outline](https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/858#issuecomment-2919637687)
  - Do the ownership change from within the driver by advertising the VOLUME_MOUNT_GROUP capability.
### Notes:
  - The changes also include secret rotation based on SecretProviderClassPodStatusStatus, but will be reverted if https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/1622 merges earlier
  - In addition, pulled up some of common repetitive code from the unit and e2e tests to make them a bit more terse
### tests added in e2e-provider:
    (leaving in the test status and runtime just for reference)
  - ok 16 Non-root POD with no FSGroup - create in 871ms
  - ok 17 Non-root POD with no FSGroup - Should fail to read non world readable secret in 186ms
  - ok 18 Non-root POD with no FSGroup - unmount succeeds in 10143ms
  - ok 19 Non-root POD with FSGroup - create in 1439ms
  - ok 20 Non-root POD with FSGroup - should read non world readable secret in 202ms
  - ok 21 Non-root POD with FSGroup - rotated secret should also be readable in 37119ms
  - ok 22 Non-root POD with FSGroup - unmount succeeds in 10177ms

### unit tests:
  - nodeserver_test
      - TestNodePublishVolume_Errors/Invalid_FSGroup
      - TestNodePublishVolume/volume_mount_with_valid_FSGroup
  - reconciler_test
      - TestReconcileError/failed_to_parse_FSGroup
      - TestReconcileNoError/reconcile_with_FSGroup
 

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [X] adds unit tests

dobsonj commented:
/ok-to-test

mytreya-rh commented:
/retest

aramase reviewed (changes requested):
The windows job failures are related to this PR.

```
E0630 16:56:06.282028   10108 atomic_writer.go:419] "unable to change file with owner" err="chown c:\\var\\lib\\kubelet\\pods\\ff425598-c3fa-480d-a6af-814831673629\\volumes\\kubernetes.io~csi\\secrets-store-inline\\mount\\..2025_06_30_16_56_06.1168464543\\secretalias: not supported by windows" logContext="secrets-store-csi-driver" fullPath="c:\\var\\lib\\kubelet\\pods\\ff425598-c3fa-480d-a6af-814831673629\\volumes\\kubernetes.io~csi\\secrets-store-inline\\mount\\..2025_06_30_16_56_06.1168464543\\secretalias" owner=-1
```

ref: https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_secrets-store-csi-driver/1841/pull-secrets-store-csi-driver-e2e-windows/1939725228552753152/artifacts/2025-06-30T170106/secrets-store.log

mytreya-rh commented:
> The windows job failures are related to this PR.
> 
> ```
> E0630 16:56:06.282028   10108 atomic_writer.go:419] "unable to change file with owner" err="chown c:\\var\\lib\\kubelet\\pods\\ff425598-c3fa-480d-a6af-814831673629\\volumes\\kubernetes.io~csi\\secrets-store-inline\\mount\\..2025_06_30_16_56_06.1168464543\\secretalias: not supported by windows" logContext="secrets-store-csi-driver" fullPath="c:\\var\\lib\\kubelet\\pods\\ff425598-c3fa-480d-a6af-814831673629\\volumes\\kubernetes.io~csi\\secrets-store-inline\\mount\\..2025_06_30_16_56_06.1168464543\\secretalias" owner=-1
> ```
> 
> ref: https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_secrets-store-csi-driver/1841/pull-secrets-store-csi-driver-e2e-windows/1939725228552753152/artifacts/2025-06-30T170106/secrets-store.log

Thanks @aramase !
Pushed a commit to [skip Chown on Windows](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/1841/files#diff-7d9dbd21a5dbf00c65daa8b70748f3e7b424b470de1ce3cb64242b39c1565229R415). Guess this is inline with FSGroup behavior as well on Windows nodes.

mytreya-rh commented:
/retest

dobsonj reviewed (commented):
I have one comment on `test/bats/e2e-provider.bats`, but otherwise LGTM. It's a useful fix, implementation looks correct, good test coverage, and passing CI tests. Netlify is warning about a line unrelated from your changes.

mytreya-rh commented:
/retest

dobsonj commented:
/lgtm

/sig storage
/triage accepted
/priority important-soon

/assign @aramase
for approval and to decide which PR should merge first between #1841 and #1622

aramase reviewed (changes requested):
First pass.

I haven't looked at the test changes yet.

mytreya-rh commented on deploy/secrets-store.csi.x-k8s.io_secretproviderclasspodstatuses.yaml:
done

mytreya-rh commented on pkg/secrets-store/nodeserver.go:
```diff
@@ -142,6 +144,21 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+		gid, err = strconv.ParseInt(fsGroupStr, 10, 64)
+		klog.V(5).Info("converted gid: %v\n", gid)
+		if err != nil {
+			klog.ErrorS(err, "failed to mount secrets store object content", "pod", klog.ObjectRef{Namespace: podNamespace, Name: podName}, "invalid FSGroup: ", fsGroupStr)
```
Done

mytreya-rh commented on pkg/secrets-store/utils.go:
```diff
@@ -99,7 +101,12 @@ func createOrUpdateSecretProviderClassPodStatus(ctx context.Context, c client.Cl
+	if gid != constants.NoGID {
+		fsGroup = strconv.FormatInt(gid, 10)
+	}
+	klog.V(5).Infof("gid string: %v for pod: %v/%v", fsGroup, namespace, podname)
```
Done

mytreya-rh commented on pkg/util/fileutil/atomic_writer.go:
```diff
@@ -410,11 +412,11 @@ func (w *AtomicWriter) writePayloadToDir(payload map[string]FileProjection, dir
 		}
 
-		if fileProjection.FsUser == nil {
+		if fileProjection.FsGroup == nil || runtimeutil.IsRuntimeWindows() {
```
Yes, as per [documentation](https://kubernetes.io/docs/concepts/windows/intro/#:~:text=If%20you%20set%20the%20.spec.os.name%20field%20to%20windows%2C%20you%20must%20not%20set%20the%20following%20fields%20in%20the%20.spec%20of%20that%20Pod%3A):


And as per code as well, the Kubernetes FSGroup implementation of writable volumes is a no-op:
[csi_mounter.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_mounter.go#L336) --> [volume_unsupported.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/volume_unsupported.go#L29)

mytreya-rh commented:
/retest

aramase commented on pkg/constants/constants.go:
```diff
@@ -0,0 +1,20 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
```suggestion
*/

```

nit: new line

aramase commented on pkg/constants/constants.go:
```diff
@@ -0,0 +1,20 @@
+package constants
+
+const (
+	NoGID = int64(-1) // Use the default gid -1 to indicate no change in FSGroup
```
```suggestion
    // NoGID is the default gid -1 to indicate no change in FSGroup
	NoGID = int64(-1)
```

aramase commented on pkg/secrets-store/nodeserver.go:
```diff
@@ -140,7 +142,22 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+		fsGroupStr := mountVol.GetVolumeMountGroup()
+		klog.V(5).Info("fsGroupStr: %v\n", fsGroupStr)
+		gid, err = strconv.ParseInt(fsGroupStr, 10, 64)
+		klog.V(5).Info("converted gid: %v\n", gid)
```
```suggestion
```

mytreya-rh replied:
done

aramase commented on pkg/secrets-store/nodeserver.go:
```diff
@@ -140,7 +142,22 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+		gid, err = strconv.ParseInt(fsGroupStr, 10, 64)
+		klog.V(5).Info("converted gid: %v\n", gid)
+		if err != nil {
+			klog.ErrorS(err, "failed to mount secrets store object content", "pod", klog.ObjectRef{Namespace: podNamespace, Name: podName}, "FSGroup: ", fsGroupStr)
```
```suggestion
			klog.ErrorS(err, "failed to mount secrets store object content", "pod", klog.ObjectRef{Namespace: podNamespace, Name: podName}, "fsGroup", fsGroupStr)
```

aramase reviewed (changes requested):
This is close.

mytreya-rh commented on pkg/secrets-store/nodeserver.go:
```diff
@@ -140,7 +142,22 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+	mountVol := req.VolumeCapability.GetMount()
+	if len(mountVol.GetVolumeMountGroup()) > 0 {
+		fsGroupStr := mountVol.GetVolumeMountGroup()
+		klog.V(5).Info("fsGroupStr: %v\n", fsGroupStr)
```
done

mytreya-rh commented on pkg/rotation/reconciler.go:
```diff
@@ -398,7 +400,18 @@ func (r *Reconciler) reconcile(ctx context.Context, spcps *secretsstorev1.Secret
-	newObjectVersions, errorReason, err := secretsstore.MountContent(ctx, providerClient, string(paramsJSON), string(secretsJSON), spcps.Status.TargetPath, string(permissionJSON), oldObjectVersions)
+	gid := constants.NoGID
+	if len(spcps.Status.FSGroup) > 0 {
+		gid, err = strconv.ParseInt(spcps.Status.FSGroup, 10, 64)
```
Thanks for the suggestion, yes, moved to [fileutil](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/1841/commits/2a1d6e01358d6e181160368472edde591577c660#diff-658fe3363d32251c5757616aec8d8baa361caa70ec7292bf1668bced7654bcd0R137-R144)

aramase reviewed (commented):
@mytreya-rh https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/1622 is merged. Please rebase this PR to resolve conflicts and squash to single commit.

mytreya-rh commented:
/retest

mytreya-rh commented:
/retest

Looks like infra issues:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_secrets-store-csi-driver/1841/pull-secrets-store-csi-driver-e2e-azure/1999365595329466368#1:build-log.txt%3A1495
> helm.go:92: 2025-12-12 06:37:38.2738949 +0000 UTC m=+0.187530371 [debug] Get "[https://sscsi-e2e--sscsi-e2e-9e5f-46678f-9lqficf8.hcp.uksouth.azmk8s.io:443/version](https://sscsi-e2e--sscsi-e2e-9e5f-46678f-9lqficf8.hcp.uksouth.azmk8s.io/version)": dial tcp: lookup sscsi-e2e--sscsi-e2e-9e5f-46678f-9lqficf8.hcp.uksouth.azmk8s.io on 172.20.0.10:53: no such host
> Kubernetes cluster unreachable

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_secrets-store-csi-driver/1841/pull-secrets-store-csi-driver-e2e-windows/1999365595853754368#1:build-log.txt%3A18
> ERROR: (Canceled) Operation is being canceled by user

mytreya-rh commented:
/retest
looks like an infra error:

> E0224 15:46:58.618759   33163 reflector.go:204] "Failed to watch" err="failed to list *unstructured.Unstructured: Get \"https://sscsi-e2e--sscsi-e2e-25c2-46678f-ync6prxe.hcp.eastus2.azmk8s.io:443/apis/batch/v1/namespaces/kube-system/jobs?fieldSelector=metadata.name%3Dsecrets-store-csi-driver-upgrade-crds&resourceVersion=4266\": dial tcp: lookup sscsi-e2e--sscsi-e2e-25c2-46678f-ync6prxe.hcp.eastus2.azmk8s.io on 172.20.0.10:53: no such host" logger="UnhandledError" reflector="k8s.io/client-go@v0.35.0/tools/cache/reflector.go:289" type="*unstructured.Unstructured"

aramase reviewed (changes requested):
Another pass.

mytreya-rh commented on test/bats/e2e-provider.bats:
```diff
@@ -25,6 +27,8 @@ export KEY_VERSION=${KEY_VERSION:-"v1"}
 # default key value returned by mock provider.
 # base64 encoded content comparision is easier in case of very long multiline string.
 export KEY_VALUE_CONTAINS=${KEY_VALUE:-"LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KVGhpcyBpcyBtb2NrIGtleQotLS0tLUVORCBQVUJMSUMgS0VZLS0tLS0K"}
+# defualt version value returned by mock provider
```
corrected all the three occurrences of this typo, thanks!

aramase commented on pkg/secrets-store/nodeserver_test.go:
```diff
@@ -57,6 +57,41 @@ func testNodeServer(t *testing.T, client client.Client, reporter StatsReporter,
+			csiPodName:            "pod1",
+			csiPodNamespace:       "default",
+			csiPodUID:             "poduid1",
+		},
```
nit: default `VolumeCapability` has nil `AccessType`. Works today because `GetMount()` on nil returns nil and `GetVolumeMountGroup()` on nil returns `""`. Fragile if we ever add a nil guard.

  ```go
                VolumeCapability: &csi.VolumeCapability{
                        AccessType: &csi.VolumeCapability_Mount{
                                Mount: &csi.VolumeCapability_MountVolume{},
                        },
                },
  ```

aramase reviewed (changes requested):
This is close.

aramase commented on pkg/util/fileutil/atomic_writer.go:
```diff
@@ -18,6 +18,9 @@ limitations under the License.
 //  * tag: v1.20.6,
 //  * commit: 8a62859e515889f07e3e3be6a1080413f17cf2c3
 //  * link: https://github.com/kubernetes/kubernetes/blob/8a62859e515889f07e3e3be6a1080413f17cf2c3/pkg/volume/util/atomic_writer.go
+// In addition, FileProjection::FSUser has been changed to FileProjection::FSGroup
```
The header comment is fine but could you also add a one-liner at the struct itself? That's where people will look when they see `FsGroup *int` and wonder why it doesn't match upstream's `FsUser *int64`:

  ```go
  // FileProjection contains file Data and access Mode.
  // FsGroup diverges from upstream's FsUser (*int64) — see file header for rationale.
  type FileProjection struct {
  ```

aramase reviewed (changes requested):
Hopefully last set of comments.

mytreya-rh commented:
/retest
seems to be an infra issue or IRSA configuration error. If this fails, will add some debug in next commit
> 0330 12:07:02.492849       1 nodeserver.go:253] "failed to mount secrets store object content" err="rpc error: code = Unknown desc = Failed to fetch secret from all regions. Verify secret exists and required permissions are granted for: SecretsManagerRotationTest-secret-d37d9220da61" pod="kube-system/basic-test-mount" isRemountRequest=false

mytreya-rh committed:
Add CSI VOLUME_MOUNT_GROUP support for FSGroup-based file ownership

Implements the CSI NodeServiceCapability RPC_VOLUME_MOUNT_GROUP so that
mounted secret files are chown'd to the pod's FSGroup.
This allows secrets to be not world-readable in non-root containers.

1. nodeServer::NodeGetCapabilities()
   advertise VOLUME_MOUNT_GROUP
2. nodeServer::NodePublishVolume()
   get POD's FSGroup if any from: req.VolumeCapability.GetMount().GetVolumeMountGroup()
3. pass the fsgroup onto (writer.go) WritePayloads()
4. include the FSGroup in the FileProjection struct (rename FileProjection::FSUser as FSGroup)
5. change AtomicWriter::writePayloadToDir() to chown the group based on FSGroup
6. Add relevant Unit tests, and e2eprovider tests
7. Bit of refactoring in the unit and e2eprovider tests to make them more terse

aramase commented on test/bats/e2e-provider.bats:
```diff
@@ -96,93 +181,83 @@ export VALIDATE_TOKENS_AUDIENCE=$(get_token_requests_audience)
+   POD_NAME="non-root-with-no-fsgroup" delete_pod
+}
+
+@test "Non-root POD with FSGroup - create" {
```
Missing one case here: **mode `0600` with `fsGroup` should still fail to read.** That's the test that proves the chown only touches the group bit and not the user bit — the exact property that would silently break if a future refactor mis-orders the args to `os.Chown(fullPath, -1, *fileProjection.FsGroup)`. Without it, swapping those args goes unnoticed: mode `0640` still works (group readable either way once you own it), and there's no negative test catching the regression.
 
 Suggested addition:
 
 ```bash
 @test "deploy e2e-provider v1 secretproviderclass crd with owner-only permissions" {
   SPC_NAME="e2e-provider-600" SECRET_MODE=0600 KEY_MODE=0600 create_spc
 }
 
 @test "Non-root POD with FSGroup - should fail to read owner-only secret" {
   SPC_NAME="e2e-provider-600" POD_NAME="non-root-with-fsgroup-600" POD_SECURITY_CONTEXT='"runAsNonRoot": true, "runAsUser": 1004, "runAsGroup": 1004, "fsGroup": 1004' create_pod
   POD_NAME="non-root-with-fsgroup-600" run read_secret
   assert_failure
 }
 
 @test "Non-root POD with FSGroup - owner-only unmount succeeds" {
   POD_NAME="non-root-with-fsgroup-600" delete_pod
 }
 ```

mytreya-rh replied:
Thanks @aramase included the above test and its [working as expected](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_secrets-store-csi-driver/1841/pull-secrets-store-csi-driver-e2e-provider-k8s-1-35-0/2064971541661093888#1:build-log.txt%3A1928).

aramase reviewed (changes requested):
@mytreya-rh just one comment and also PTAL at the copilot review comment.

mytreya-rh commented on test/bats/e2e-provider.bats:
```diff
@@ -33,11 +37,92 @@ export LABEL_VALUE=${LABEL_VALUE:-"test"}
+}
+
+function disable_secret_rotation() {
+  local curl_pod_name=$1
+  local pod_ip=$(kubectl get pod -n kube-system -l app=csi-secrets-store-e2e-provider -o jsonpath="{.items[0].status.podIP}")  
+  run kubectl exec ${curl_pod_name} -n rotation -- curl http://${pod_ip}:8080/rotation?rotated=false
+}
```
Thanks for the catch, @copilot
The pod creation is now moved to setup_file and the same POD will be reused across the various rotation tests. The pod will be deleted in teardown when the namespace gets deleted.
This way we wont create and delete the POD multiple times.
Instead of run kubectl, the disable_secret_rotation function just executes kubectl directly to not let the error go unnoticed.

Nemric commented:
Hi, what's the status of this PR that seems to help a lot for files that need 600 ownership

mytreya-rh commented:
> Hi, what's the status of this PR that seems to help a lot for files that need 600 ownership

hi @Nemric This PR helps non-root containers mount and use non world-readable secrets.
Thus, making it possible to mount secrets with 440 permissions instead of the default 666.
This in itself improves the security posture of the mounted secrets and also the PODs using them.
The way it works is by respecting the FSGroup field of the POD spec.

But, as of now, the FSUser which is derived from the runAsUser field of the POD/Container spec is [NOT passed to CSIDrivers](https://github.com/kubernetes/kubernetes/blob/a818af18fe29d999d6741234c8cd72709ef2f424/pkg/volume/volume.go#L119-L120) and changing the USER ownership will require changes in kubelet and perhaps in the NodePublishVolume API.
We can discuss this as a follow up as it would be a useful feature. WDYT @aramase ?

Nemric commented:
would that help ? https://github.com/kubernetes/kubernetes/pull/139764
[edit] : https://github.com/kubernetes/enhancements/issues/5936

mytreya-rh commented:
> would that help ? [kubernetes/kubernetes#139764](https://github.com/kubernetes/kubernetes/pull/139764)

surely.. thanks for sharing

ryan0x44 commented:
Would be great to see this merged, it looks ready to go - but let me know if I can help with anything!
